compiler/src/dotty/tools/MainGenericRunner.scala

```diff
@@ -0,0 +1,154 @@
+    this.copy(scriptArgs = scriptArgs.appendedAll(args.toList))
+
+  def withTargetScript(file: String): Settings =
+    Try(Source.fromFile(file)).toOption match
+      case Some(_) => this.copy(targetScript = file)
+      case None      =>
+        println(s"not found $file")
+        this.copy(exitCode = 2)
+  end withTargetScript
+
+  def withSave: Settings =
+    this.copy(save = true)
+
+  def withModeShouldBeRun: Settings =
+    this.copy(modeShouldBeRun = true)
+}
+
+object MainGenericRunner {
+
+  val classpathSeparator = File.pathSeparator
+
+  @sharable val javaOption = raw"""-J(.*)""".r
+
+  @tailrec
+  def process(args: List[String], settings: Settings): Settings = args match
+    case Nil =>
+      settings
+    case "-run" :: tail =>
+      process(tail, settings.withExecuteMode(ExecuteMode.Run))
+    case ("-cp" | "-classpath" | "--class-path") :: cp :: tail =>
+      process(tail, settings.copy(classPath = settings.classPath.appended(cp)))
+    case ("-version" | "--version") :: _ =>
+      settings.copy(
+        executeMode = ExecuteMode.Repl,
+        residualArgs = List("-version")
+      )
+    case ("-v" | "-verbose" | "--verbose") :: tail =>
+      process(
+        tail,
+        settings.copy(
+          verbose = true,
+          residualArgs = settings.residualArgs :+ "-verbose"
+        )
+      )
+    case "-save" :: tail =>
+      process(tail, settings.withSave)
+    case (o @ javaOption(striped)) :: tail =>
+      process(tail, settings.withJavaArgs(striped).withScalaArgs(o))
+    case arg :: tail =>
+      val line = Try(Source.fromFile(arg).getLines.toList).toOption.flatMap(_.headOption)
+      if arg.endsWith(".scala") || arg.endsWith(".sc") || (line.nonEmpty && raw"#!.*scala".r.matches(line.get)) then
+        settings
+          .withExecuteMode(ExecuteMode.Script)
+          .withTargetScript(arg)
+          .withScriptArgs(tail*)
+      else
+        val newSettings = if arg.startsWith("-") then settings else settings.withModeShouldBeRun
+        process(tail, newSettings.withResidualArgs(arg))
+
+  def main(args: Array[String]): Unit =
+    val settings = process(args.toList, Settings())
+    if settings.exitCode != 0 then System.exit(settings.exitCode)
+
+    def run(mode: ExecuteMode): Unit = mode match
+      case ExecuteMode.Repl =>
+        val properArgs =
+          List("-classpath", settings.classPath.mkString(classpathSeparator)).filter(Function.const(settings.classPath.nonEmpty))
+            ++ settings.residualArgs
+        repl.Main.main(properArgs.toArray)
+      case ExecuteMode.Run =>
+        val properArgs =
+          val newClasspath = settings.classPath ++ getClasspath :+ "."
+          List("-classpath", newClasspath.mkString(classpathSeparator)).filter(Function.const(newClasspath.nonEmpty))
+            ++ settings.residualArgs
+        s"java ${settings.javaArgs.mkString(" ")} ${properArgs.mkString(" ")}".! // For now we collect classpath that coursier provides for convenience
```

Comment:
No, that's not what we want to do here: we don't want to fork another JVM - we want to launch the main in a new classloader constructed with the given classpath.  In scala/scala that's `ObjectRunner.runAndCatch` if you want to mirror what's being done there.

Reply:
Thanks for the advice. I couldn't find proper entry from scala sources and didn't think about launching the main in a new classloader. Will change